pdns/lua-record.cc

```diff
@@ -60,30 +60,145 @@ class IsUpOracle
+    /* first check ? */
+    std::atomic<bool> first{true};
+    /* last time the status was accessed */
+    std::atomic<time_t> lastAccess{0};
+  };
+
+  pthread_rwlock_t d_lock;
 public:
+  IsUpOracle()
+  {
+    pthread_rwlock_init(&d_lock, nullptr);
+  }
+  ~IsUpOracle()
+  {
+    pthread_rwlock_destroy(&d_lock);
+  }
   bool isUp(const ComboAddress& remote, const opts_t& opts);
   bool isUp(const ComboAddress& remote, const std::string& url, const opts_t& opts);
   bool isUp(const CheckDesc& cd);
 
 private:
-  void checkURLThread(ComboAddress rem, std::string url, const opts_t& opts);
-  void checkTCPThread(ComboAddress rem, const opts_t& opts);
+  void checkURL(const CheckDesc& cd, const bool status, const bool first = false)
+  {
+    try {
+      int timeout = 1;
+      if (cd.opts.count("timeout")) {
+        timeout = std::atoi(cd.opts.at("timeout").c_str());
+      }
+      string useragent = productName();
+      if (cd.opts.count("useragent")) {
+        useragent = cd.opts.at("useragent");
+      }
+      MiniCurl mc(useragent);
 
-  struct Checker
+      string content;
+      if (cd.opts.count("source")) {
+        ComboAddress src(cd.opts.at("source"));
+        content=mc.getURL(cd.url, &cd.rem, &src, timeout);
+      }
+      else {
+        content=mc.getURL(cd.url, &cd.rem, nullptr, timeout);
+      }
+      if (cd.opts.count("stringmatch") && content.find(cd.opts.at("stringmatch")) == string::npos) {
+        throw std::runtime_error(boost::str(boost::format("unable to match content with `%s`") % cd.opts.at("stringmatch")));
+      }
+      if(!status) {
+        g_log<<Logger::Warning<<"LUA record monitoring declaring "<<cd.rem.toString()<<" UP for URL "<<cd.url<<"!"<<endl;
+      }
+      setUp(cd);
+    }
+    catch(std::exception& ne) {
+      if(status || first)
+        g_log<<Logger::Warning<<"LUA record monitoring declaring "<<cd.rem.toString()<<" DOWN for URL "<<cd.url<<", error: "<<ne.what()<<endl;
+      setDown(cd);
+    }
+  }
+  void checkTCP(const CheckDesc& cd, const bool status, const bool first = false) {
+    try {
+      int timeout = 1;
+      if (cd.opts.count("timeout")) {
+        timeout = std::atoi(cd.opts.at("timeout").c_str());
+      }
+      Socket s(cd.rem.sin4.sin_family, SOCK_STREAM);
+      ComboAddress src;
+      s.setNonBlocking();
+      if (cd.opts.count("source")) {
+        src = ComboAddress(cd.opts.at("source"));
+        s.bind(src);
+      }
+      s.connect(cd.rem, timeout);
+      if (!status) {
+        g_log<<Logger::Warning<<"Lua record monitoring declaring TCP/IP "<<cd.rem.toStringWithPort()<<" ";
+        if(cd.opts.count("source"))
+          g_log<<"(source "<<src.toString()<<") ";
+        g_log<<"UP!"<<endl;
+      }
+      setUp(cd);
+    }
+    catch (const NetworkError& ne) {
+      if(status || first) {
+        g_log<<Logger::Warning<<"Lua record monitoring declaring TCP/IP "<<cd.rem.toStringWithPort()<<" DOWN: "<<ne.what()<<endl;
+      }
+      setDown(cd);
```

Comment:
I wonder if we should catch some other exception types here. I'm for example thinking of an invalid "source" address throwing a `PDNSException`.